kubernetes/config/kube_config.c

```diff
@@ -0,0 +1,309 @@
+        }
+    }
+
+    sslConfig_t *sc = sslConfig_create(client_cert_file, client_key_file, ca_file, insecure_skip_tls_verify);
+    if (sc) {
+        *pSslConfig = sc;
+    } else {
+        rc = -1;
+    }
+
+    if (client_cert_file) {
+        free(client_cert_file);
+    }
+    if (client_key_file) {
+        free(client_key_file);
+    }
+    if (ca_file) {
+        free(ca_file);
+    }
+
+    return rc;
+}
+
+static int setApiKeys(list_t ** pApiKeys, kubeconfig_user_t * user)
+{
+    int rc = 0;
+    list_t *apiKeys = list_create();
+    if (apiKeys) {
+
+        /* under development for the token based authentication */
+
+        *pApiKeys = apiKeys;
+    } else {
+        rc = -1;
+    }
+    return rc;
+}
+
+static char *getWorkingConfigFile(const char *configFileNamePassedIn)
+{
+    char *configFileName = NULL;
+    const char *kubeconfig_env = NULL;
+    const char *homedir_env = NULL;
+
+    if (configFileNamePassedIn) {
+        configFileName = strdup(configFileNamePassedIn);
+    } else {
+        kubeconfig_env = getenv(ENV_KUBECONFIGE);
+        if (kubeconfig_env) {
+            configFileName = strdup(kubeconfig_env);
+        } else {
+            homedir_env = getenv(ENV_HOME);
+            if (homedir_env) {
+                configFileName = calloc(strlen(homedir_env) + strlen(KUBE_CONFIG_DEFAULT_LOCATION) + 1, sizeof(char));
+                if (configFileName) {
+                    sprintf(configFileName, KUBE_CONFIG_DEFAULT_LOCATION, homedir_env);
+                }
+            }
+        }
+    }
+
+    return configFileName;
+}
+
+static kubeconfig_context_t *kubeconfig_get_current_context(kubeconfig_context_t ** contexts, int contexts_count, const char *current_context_str)
+{
+    kubeconfig_context_t *current_context = NULL;
+
+    if (NULL == contexts || NULL == current_context_str) {
+        return NULL;
+    }
+
+    int i = 0;
+    for (i = 0; i < contexts_count; i++) {
+        if (0 == strcmp(contexts[i]->name, current_context_str)) {
+            current_context = contexts[i];
+            break;
+        }
+    }
+
+    return current_context;
+}
+
+static kubeconfig_cluster_t *kubeconfig_get_current_cluster(kubeconfig_cluster_t ** clusters, int clusters_count, const char *cluster_name)
+{
+    kubeconfig_cluster_t *current_cluster = NULL;
+
+    if (NULL == clusters || NULL == cluster_name) {
+        return NULL;
+    }
+
+    int i = 0;
+    for (i = 0; i < clusters_count; i++) {
+        if (0 == strcmp(clusters[i]->name, cluster_name)) {
+            current_cluster = clusters[i];
+            break;
+        }
+    }
+
+    return current_cluster;
+}
+
+static kubeconfig_user_t *kubeconfig_get_current_user(kubeconfig_user_t ** users, int users_count, const char *user_name)
+{
+    kubeconfig_user_t *current_user = NULL;
+
+    if (NULL == users || NULL == user_name) {
+        return NULL;
+    }
+
+    int i = 0;
+    for (i = 0; i < users_count; i++) {
```

Comment:
optional suggestion:

I feel like if you made the objects union types you might be able to eliminate all of this redundant (but slightly different) code for each of these types.

Reply:
It's a very good idea. I ever want to use Macro to reduce the reduncant code. But obviously, union is better.

Reply:
Hi @brendandburns 

Is the below design same with your thinking of union ?

```c
typedef struct kubeconfig_t {
    char *fileName;
    char *apiVersion;
    char *preferences;
    char *kind;
    char *current_context;
    kubeconfig_property_t **contexts;
    int  contexts_count;
    kubeconfig_property_t **clusters;
    int  clusters_count;
    kubeconfig_property_t **users;
    int  users_count;
} kubeconfig_t;

typedef struct kubeconfig_property_t {
    char *name;
    union {
        kubeconfig_cluster_t cluster;
        kubeconfig_user_t user;
        kubeconfig_context_t context;
    };
} kubeconfig_property_t;

typedef struct kubeconfig_cluster_t {
    char *server;
    char *certificate_authority_data;
} kubeconfig_cluster_t ;

typedef struct kubeconfig_context_t {
    char *cluster;
    char *user;
} kubeconfig_context_t ;

typedef struct kubeconfig_user_t {
    char *client_certificate_data;
    char *client_key_data;
    int insecure_skip_tls_verify;
...
} kubeconfig_user_t;
```

Reply:
Yeah, that looks about like what I was thinking about.